### ISSUE TITLE
[shopify-api] GraphQL request with custom shop domain is not working

### ISSUE THREAD

**Issue by McTom234** [Shopify/shopify-app-js#1868]
# Issue summary

Before opening this issue, I have:

- [X] Upgraded to the latest version of the relevant packages
  - `@shopify/*` package and version:
  - Node version: 20.11.1
  - Operating system: Linux
- [X] Set `{ logger: { level: LogSeverity.Debug } }` in my configuration, when applicable
- [X] Found a reliable way to reproduce the problem that indicates it's a problem with the package
- [X] Looked for similar issues in this repository
- [ ] Checked that this isn't an issue with a Shopify API
  - If it is, please create a post in the [Shopify community forums](https://community.shopify.com/c/partners-and-developers/ct-p/appdev) or report it to [Shopify Partner Support](https://help.shopify.com/en/support/partners/org-select)

TL;DR: custom shop domains cannot be used with this library, due to a wrong redirect handling.

Prerequisites:
- shop `A` has a custom domain `a.com`
- you have valid credentials for shop `A` to make API requests (tested by using the `Location` header of the redirect response)

The request to the GraphQL endpoint for shop `A` on `a.com/...` returns a `301` response and redirects to the shop's `{shop}.myshopify.com` host. The default redirect handling of a `301` response in node-fetch is to set the request method to `GET`. That results in a `404` response on the API (which I think is not optimal - see 405; but this is an API, not library issue).

## Expected behavior

- The Shopify API should return a `308` for this request.
- The Shopify API client should handle the redirect for GraphQL requests to use a `POST` request again.

## Actual behavior

- The Shopify API returns a `301`, which resets the request method to `GET`, as this is the `node-fetch` default.

## Steps to reproduce the problem

_I'd say, the explanation does a quite good job for this..._

## Other

This is partly a Shopify API issue. But the library should be able to handle this case (imo).

**Comment by McTom234** [comment]
Dear maintainers, is it worth working on the other clients or would you reject the PR even if the tests weren't failing?

**Comment by McTom234** [comment]
Hi @lizkenyon,

That is what I see as well, the underlying issue is with the API itself. It is possible to make requests from a custom domain, but they get redirected to the myshopify domain. The redirect response code is `301`, which causes the fetch implementation to continue with `GET` requests.

The use case is a situation where only the custom domain is provided. Users enter their shop domain and some users enter the custom domain. Since it is a basic redirect, it shouldn't be a big issue.

Additionally, I want to point to the [RFC2616](https://datatracker.ietf.org/doc/html/rfc2616#section-10.3.2). It clearly states that the `POST` to `GET` conversion is an error. Furthermore, it also says that the `POST` request shouldn't be forwarded at all. This brings up the discussion whether the SDK implementation is wrong or the API itself, for which I say, the API implementation is wrong. But there is no way to contact someone about this, so I opened this PR. If you can forward this issue to the correct teams, this PR never needs to be merged. If the change is not possible, there should be a dirty workaround like this or a domain check when making requests.

The current implementation can cause much confusion, since it converts requests to GET requests, which is not correct for the GraphQL API.

**Comment by lizkenyon** [comment]
Thanks for the quick reply. I am in contact with the API team. 

> The use case is a situation where only the custom domain is provided. Users enter their shop domain and some users enter the custom domain

Could you explain more about this?  What is the scenario where user's are entering their shop domain? Are you building public or custom apps?

**Comment by lizkenyon** [comment]
Thank you @McTom234 

The API team wants to better understand your use case. 
Sorry I just to clarify some terms to ensure we are on the same page.

> We are building a custom app. Currently, we are in the review process

Could you clarify your [distribution method](https://shopify.dev/docs/apps/launch/distribution/select-distribution-method)?
Public or Custom?

When you say "review process" you are referring to an internal review? Or the Shopify App Review?

**Comment by lizkenyon** [comment]
Could you confirm what is the use case where manual entry of domains is required? 

The GraphQL team want to understand the use case so why this is necessary so they can best prioritize the resolution.

**Comment by McTom234** [comment]
Sorry for the late reply. We have no active use case anymore.

Besides that, the response code is still wrong according to the RFC.

**Comment by byrichardpowell** [comment]
Since there isn't a use case anymore.  I'm going to close this issue
